### PR TITLE
Shut down the daemon after 1h of inactivity

### PR DESCRIPTION
**Pull request by hughsie**
Plugins are allowed to 'opt-out' of this behaviour using _RULE_INHIBITS_IDLE.
This should be used where waking up the hardware to coldplug is expensive,
either from a power consumption point of view, or if other artifacts are going
to be seem -- for instance if the screen flickers when probing display devices.

This functionality is also inhibited when the actual upgrade is happening,
for obvious reasons.

Admins can turn off this auto-sleep behaviour by editing the daemon.conf file.

This needs more work, for instance adding hardware hotplug auto-start of the
daemon, for instance:

SUBSYSTEM=="usb", TAG+="systemd", ENV{SYSTEMD_WANTS}+="fwupd.target"
then /usr/lib/systemd/system/fwupd.service is WantedBy=fwupd.target

Fixes some of https://github.com/hughsie/fwupd/issues/417

### CODE REVIEW

File: src/fu-engine.c [hunk 1]
```diff
@@ -88,6 +90,7 @@ static void
 fu_engine_emit_changed (FuEngine *self)
 {
 	g_signal_emit (self, signals[SIGNAL_CHANGED], 0);
+	fu_idle_reset (self->idle);
```
Comment (superm1) [inline src/fu-engine.c]: Just call `fu_engine_idle_reset` here with the self argument instead?

**Comment by superm1** [comment]
Can you add inhibitor to the synaptics plugin too?

**Comment by hughsie** [comment]
> reset engine idle in fwupdtool at startup

I don't think it'll run for fwupdtool at all; there is no mainloop to quit, and it only sets the timeout to nonzero if FU_APP_FLAGS_NO_IDLE_SOURCES isn't set, which is set in fwupdtool.

**Comment by hughsie** [comment]
> Can you add inhibitor to the synaptics plugin too?

With a reason like `SynapticsMST can cause the screen to flash when probing`?

**Comment by superm1** [comment]
Yeah

**Comment by hughsie** [comment]
@gicmo I'd be interested in what you think of this too. I'm tempted to merge it without the systemd wakeup bits; if you insert a USB device with a pending firmware update do we have to notify straight away?

On the desktop gnome-software is polling fwupd at least every hour so if it's something like a keyboard you leave plugged in then we're going to show the right thing eventually. If you're usig the CLI like on a raspberry-pi we don't have any way to notify the user as there's no "client" running.

**Comment by superm1** [comment]
FWIW I think you're right on the lack of wake-up bits for most contexts.  Systems using gnome-software and KDE discover it's a slam dunk.

I think it would be useful however to ship stuff udev rules and systemd changes (if needed) that are disabled by default but can be used in case of working on a frontend-less installation.

The main place I'm thinking this would make sense is in a environment where you might not have a traditional frontend checking for updates constantly but instead have a management agent that might be waiting for `DeviceAdded` dbus events that a new device was plugged in to decide to start up an updater process and check.

This is how I (currently) envision this fitting into ChromeOS and other environments without a frontend typically running (@SaludLemus maybe you can confirm if you guys think otherwise).

**Comment by hughsie** [comment]
> but instead have a management agent that might be waiting for DeviceAdded dbus events

I'd like to sit down and actually design something that's suitable; rather than guessing what works.

> This is how I (currently) envision this fitting into ChromeOS and other environments without a frontend 

I'd love to sit down with a few of the ChromeOS developers and actually talk about how they want to use fwupd in an ideal world; I think it would be really useful for both sides.

**Comment by superm1** [comment]
Makes sense to me.  Right now I understand that they want to use it non-interactively, but some of this will be influenced by results of some in progress plugin improvements.  Once that stuff is further along I'll try to arrange some time to sit down together.

**Pull request merged**